// ng6-notify-popup/esm5/lib/ng6-notify-popup.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { Ng6NotifyPopupComponent } from './ng6-notify-popup.component';
import { CommonModule } from '@angular/common';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { ComponentInjectService } from './services/component-inject.service';
import { Ng6NotifyPopupService } from './services/ng6-notify-popup.service';
var Ng6NotifyPopupModule = /** @class */ (function () {
    function Ng6NotifyPopupModule() {
    }
    /**
     * @return {?}
     */
    Ng6NotifyPopupModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: Ng6NotifyPopupModule,
            providers: [
                Ng6NotifyPopupService,
                ComponentInjectService
            ]
        };
    };
    Ng6NotifyPopupModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        BrowserAnimationsModule
                    ],
                    declarations: [
                        Ng6NotifyPopupComponent
                    ],
                    providers: [
                        ComponentInjectService
                    ],
                    entryComponents: [
                        Ng6NotifyPopupComponent
                    ]
                },] },
    ];
    return Ng6NotifyPopupModule;
}());
export { Ng6NotifyPopupModule };
