// ng6-notify-popup/esm5/lib/ng6-notify-popup.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { animate, state, style, transition, trigger } from '@angular/animations';
var Ng6NotifyPopupComponent = /** @class */ (function () {
    function Ng6NotifyPopupComponent() {
        this.destroyComponent = new EventEmitter();
    }
    /**
     * @return {?}
     */
    Ng6NotifyPopupComponent.prototype.dismiss = /**
     * @return {?}
     */
    function () {
        this.destroyComponent.emit(true);
    };
    Ng6NotifyPopupComponent.decorators = [
        { type: Component, args: [{
                    selector: 'notify',
                    template: "\n    <div [@shrinkOut]=\"fade\" class=\"trb-notify {{'trb-' + type}}\"\n         [ngClass]=\"{'trb-top': position=='top','trb-bottom':position=='bottom','trb-fixed':location=='body','trb-absolute':location!='body','trb-sticky':sticky==true}\">\n      <span class=\"trb-message\">{{notifyText}}</span>\n      <span class=\"trb-dismiss\" (click)=\"dismiss()\">\u00D7</span>\n    </div>",
                    styles: [".trb-notify{color:#fff;cursor:default;font-size:1.3em;left:0;opacity:0;padding:25px 80px;right:0;display:block;text-align:center;user-select:none;z-index:9999;-webkit-touch-callout:none;-webkit-user-select:none;-khtml-user-select:none;-moz-user-select:none;-ms-user-select:none}.trb-component{position:absolute}.trb-top{top:0}.trb-bottom{bottom:0}.trb-fixed{position:fixed}.trb-absolute{position:absolute}.trb-dismiss{background-color:#333;border-radius:15px;box-shadow:inset 2px 2px 7px 2px #000;color:#ddd;cursor:pointer;display:none;font-size:1.25em;font-weight:700;height:30px;line-height:30px;opacity:.2;position:absolute;right:40px;text-shadow:1px 1px 5px #000;top:25px;width:30px}.trb-sticky .trb-dismiss{display:block}.trb-dismiss:hover{background-color:#000}.trb-dismiss:active{background-color:#666}@media only screen and (max-width:480px){.trb-notify{font-size:1em;padding:12px 25px}.trb-dismiss{font-size:1em;height:20px;line-height:20px;right:5px;top:5px;width:20px}}.trb-info{background-color:#0e90d2}.trb-error{background-color:#dd514c}.trb-success{background-color:#5eb95e}.trb-warn{background-color:#f37b1d}.trb-grimace{background-color:#8058a5}.trb-prime.trb-info{background-color:#03c}.trb-prime.trb-error{background-color:red}.trb-prime.trb-success{background-color:#0c0}.trb-prime.trb-warn{background-color:#f90}.trb-prime.trb-grimace{background-color:#609}.trb-pastel.trb-info{background-color:#7ea7d8}.trb-pastel.trb-error{background-color:#f6989d}.trb-pastel.trb-success{background-color:#82ca9d}.trb-pastel.trb-warn{background-color:#fdc68a}.trb-pastel.trb-grimace{background-color:#a187be}.trb-pitchy.trb-info{background-color:#003471}.trb-pitchy.trb-error{background-color:#9e0b0f}.trb-pitchy.trb-success{background-color:#007236}.trb-pitchy.trb-warn{background-color:#a36209}.trb-pitchy.trb-grimace{background-color:#440e62}"],
                    animations: [
                        trigger('shrinkOut', [
                            state('show', style({
                                opacity: 1
                            })),
                            state('hide', style({
                                opacity: 0
                            })),
                            transition('show => hide', animate('700ms ease-out')),
                            transition('* => show', animate('700ms ease-out'))
                        ])
                    ]
                },] },
    ];
    Ng6NotifyPopupComponent.propDecorators = {
        position: [{ type: Input }],
        duration: [{ type: Input }],
        type: [{ type: Input }],
        notifyText: [{ type: Input }],
        fade: [{ type: Input }],
        sticky: [{ type: Input }],
        location: [{ type: Input }],
        destroyComponent: [{ type: Output }]
    };
    return Ng6NotifyPopupComponent;
}());
export { Ng6NotifyPopupComponent };
if (false) {
    /** @type {?} */
    Ng6NotifyPopupComponent.prototype.position;
    /** @type {?} */
    Ng6NotifyPopupComponent.prototype.duration;
    /** @type {?} */
    Ng6NotifyPopupComponent.prototype.type;
    /** @type {?} */
    Ng6NotifyPopupComponent.prototype.notifyText;
    /** @type {?} */
    Ng6NotifyPopupComponent.prototype.fade;
    /** @type {?} */
    Ng6NotifyPopupComponent.prototype.sticky;
    /** @type {?} */
    Ng6NotifyPopupComponent.prototype.location;
    /** @type {?} */
    Ng6NotifyPopupComponent.prototype.destroyComponent;
}
