// ng6-notify-popup/esm5/lib/services/ng6-notify-popup.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { NotificationConfig } from '../notificationConfig';
import { Ng6NotifyPopupComponent } from '../ng6-notify-popup.component';
import { ComponentInjectService } from './component-inject.service';
import * as i0 from "@angular/core";
import * as i1 from "./component-inject.service";
var Ng6NotifyPopupService = /** @class */ (function () {
    function Ng6NotifyPopupService(compInject) {
        this.compInject = compInject;
        this.defaultConfigVar = new NotificationConfig('bottom', 3000, 'error', 'body', false, '');
    }
    /**
     * @param {?} configObject
     * @return {?}
     */
    Ng6NotifyPopupService.prototype.defaultConfig = /**
     * @param {?} configObject
     * @return {?}
     */
    function (configObject) {
        this.setConfig(configObject, this.defaultConfigVar);
    };
    /**
     * @return {?}
     */
    Ng6NotifyPopupService.prototype.destroy = /**
     * @return {?}
     */
    function () {
        clearTimeout(this.cleartime);
        clearTimeout(this.clearhide);
        if (this.activeNotificationCompRef) {
            this.compInject.destroyComponent(this.activeNotificationCompRef);
            this.activeNotificationCompRef = null;
        }
    };
    /**
     * @param {?} notificationText
     * @param {?=} notificationConfig
     * @return {?}
     */
    Ng6NotifyPopupService.prototype.show = /**
     * @param {?} notificationText
     * @param {?=} notificationConfig
     * @return {?}
     */
    function (notificationText, notificationConfig) {
        var _this = this;
        this.destroy();
        /** @type {?} */
        var config = new NotificationConfig(this.defaultConfigVar.position, this.defaultConfigVar.duration, this.defaultConfigVar.type, this.defaultConfigVar.location, this.defaultConfigVar.sticky, notificationText);
        if (notificationConfig != undefined && notificationConfig != null)
            this.setConfig(notificationConfig, config);
        if (config.location == 'body')
            this.activeNotificationCompRef = this.compInject.appendComponentToBody(Ng6NotifyPopupComponent, config, document.querySelector('body'));
        else
            this.activeNotificationCompRef = this.compInject.appendComponent(Ng6NotifyPopupComponent, config, document.querySelector(config.location));
        this.activeNotificationCompRef.instance.fade = 'show';
        if (!this.activeNotificationCompRef.instance.sticky) {
            this.cleartime = window.setTimeout(function () {
                _this.activeNotificationCompRef.instance.fade = 'hide';
                _this.clearhide = window.setTimeout(function () {
                    _this.destroy();
                }, 700);
            }, config.duration);
        }
        else
            this.activeNotificationCompRef.instance.destroyComponent.subscribe(function (value) {
                _this.activeNotificationCompRef.instance.fade = 'hide';
                _this.clearhide = window.setTimeout(function () {
                    _this.destroy();
                }, 700);
            });
    };
    /**
     * @param {?} configObject
     * @param {?} targetObject
     * @return {?}
     */
    Ng6NotifyPopupService.prototype.setConfig = /**
     * @param {?} configObject
     * @param {?} targetObject
     * @return {?}
     */
    function (configObject, targetObject) {
        /** @type {?} */
        var props = Object.getOwnPropertyNames(configObject);
        try {
            for (var props_1 = tslib_1.__values(props), props_1_1 = props_1.next(); !props_1_1.done; props_1_1 = props_1.next()) {
                var prop = props_1_1.value;
                (/** @type {?} */ (targetObject))[prop] = (/** @type {?} */ (configObject))[prop];
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (props_1_1 && !props_1_1.done && (_a = props_1.return)) _a.call(props_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        var e_1, _a;
    };
    Ng6NotifyPopupService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] },
    ];
    /** @nocollapse */
    Ng6NotifyPopupService.ctorParameters = function () { return [
        { type: ComponentInjectService }
    ]; };
    /** @nocollapse */ Ng6NotifyPopupService.ngInjectableDef = i0.defineInjectable({ factory: function Ng6NotifyPopupService_Factory() { return new Ng6NotifyPopupService(i0.inject(i1.ComponentInjectService)); }, token: Ng6NotifyPopupService, providedIn: "root" });
    return Ng6NotifyPopupService;
}());
export { Ng6NotifyPopupService };
if (false) {
    /** @type {?} */
    Ng6NotifyPopupService.prototype.defaultConfigVar;
    /** @type {?} */
    Ng6NotifyPopupService.prototype.activeNotificationCompRef;
    /** @type {?} */
    Ng6NotifyPopupService.prototype.cleartime;
    /** @type {?} */
    Ng6NotifyPopupService.prototype.clearhide;
    /** @type {?} */
    Ng6NotifyPopupService.prototype.compInject;
}
