// ng6-notify-popup/esm2015/lib/services/ng6-notify-popup.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { NotificationConfig } from '../notificationConfig';
import { Ng6NotifyPopupComponent } from '../ng6-notify-popup.component';
import { ComponentInjectService } from './component-inject.service';
import * as i0 from "@angular/core";
import * as i1 from "./component-inject.service";
export class Ng6NotifyPopupService {
    /**
     * @param {?} compInject
     */
    constructor(compInject) {
        this.compInject = compInject;
        this.defaultConfigVar = new NotificationConfig('bottom', 3000, 'error', 'body', false, '');
    }
    /**
     * @param {?} configObject
     * @return {?}
     */
    defaultConfig(configObject) {
        this.setConfig(configObject, this.defaultConfigVar);
    }
    /**
     * @return {?}
     */
    destroy() {
        clearTimeout(this.cleartime);
        clearTimeout(this.clearhide);
        if (this.activeNotificationCompRef) {
            this.compInject.destroyComponent(this.activeNotificationCompRef);
            this.activeNotificationCompRef = null;
        }
    }
    /**
     * @param {?} notificationText
     * @param {?=} notificationConfig
     * @return {?}
     */
    show(notificationText, notificationConfig) {
        this.destroy();
        /** @type {?} */
        let config = new NotificationConfig(this.defaultConfigVar.position, this.defaultConfigVar.duration, this.defaultConfigVar.type, this.defaultConfigVar.location, this.defaultConfigVar.sticky, notificationText);
        if (notificationConfig != undefined && notificationConfig != null)
            this.setConfig(notificationConfig, config);
        if (config.location == 'body')
            this.activeNotificationCompRef = this.compInject.appendComponentToBody(Ng6NotifyPopupComponent, config, document.querySelector('body'));
        else
            this.activeNotificationCompRef = this.compInject.appendComponent(Ng6NotifyPopupComponent, config, document.querySelector(config.location));
        this.activeNotificationCompRef.instance.fade = 'show';
        if (!this.activeNotificationCompRef.instance.sticky) {
            this.cleartime = window.setTimeout(() => {
                this.activeNotificationCompRef.instance.fade = 'hide';
                this.clearhide = window.setTimeout(() => {
                    this.destroy();
                }, 700);
            }, config.duration);
        }
        else
            this.activeNotificationCompRef.instance.destroyComponent.subscribe((value) => {
                this.activeNotificationCompRef.instance.fade = 'hide';
                this.clearhide = window.setTimeout(() => {
                    this.destroy();
                }, 700);
            });
    }
    /**
     * @param {?} configObject
     * @param {?} targetObject
     * @return {?}
     */
    setConfig(configObject, targetObject) {
        /** @type {?} */
        const props = Object.getOwnPropertyNames(configObject);
        for (const prop of props) {
            (/** @type {?} */ (targetObject))[prop] = (/** @type {?} */ (configObject))[prop];
        }
    }
}
Ng6NotifyPopupService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] },
];
/** @nocollapse */
Ng6NotifyPopupService.ctorParameters = () => [
    { type: ComponentInjectService }
];
/** @nocollapse */ Ng6NotifyPopupService.ngInjectableDef = i0.defineInjectable({ factory: function Ng6NotifyPopupService_Factory() { return new Ng6NotifyPopupService(i0.inject(i1.ComponentInjectService)); }, token: Ng6NotifyPopupService, providedIn: "root" });
if (false) {
    /** @type {?} */
    Ng6NotifyPopupService.prototype.defaultConfigVar;
    /** @type {?} */
    Ng6NotifyPopupService.prototype.activeNotificationCompRef;
    /** @type {?} */
    Ng6NotifyPopupService.prototype.cleartime;
    /** @type {?} */
    Ng6NotifyPopupService.prototype.clearhide;
    /** @type {?} */
    Ng6NotifyPopupService.prototype.compInject;
}
